src/runtime/virtcontainers/pkg/cgroups/cgroups.go

```diff
@@ -116,13 +116,18 @@ func sandboxDevices() []specs.LinuxDeviceCgroup {
 
 func NewCgroup(path string, resources *specs.LinuxResources) (*Cgroup, error) {
 	var err error
+	var cgroup cgroups.Cgroup
 
 	cgroupPath, err := ValidCgroupPath(path, IsSystemdCgroup(path))
 	if err != nil {
 		return nil, err
 	}
 
-	cgroup, err := cgroups.New(cgroups.V1, cgroups.StaticPath(cgroupPath), resources)
+	if !IsSystemdCgroup(path) {
+		cgroup, err = cgroups.New(cgroups.V1, cgroups.StaticPath(cgroupPath), resources)
+	} else {
+		cgroup, err = cgroups.New(cgroups.Systemd, cgroups.Slice(getSliceName(cgroupPath), getUnitName(cgroupPath)), resources)
```

Comment:
It would be cleaner to factorize the `Hierarchy` and `path` building part into a common routine. Something from `pkg/cgroups/utils.go` like:

```go
func cgroupHierarchy(path string) (cgroups.Hierarchy, cgroups.Path) {
         if !IsSystemdCgroup(path) {
                 return cgroups.V1, cgroups.StaticPath(path)
         } else {
                 return cgroups.Systemd, cgroups.Slice(getSliceName(path), getUnitName(path))
         }
}
```

and then the above code would become simpler to read and maintain:

```go
     cgroup, err := cgroups.New(cgroupHierarchy(cgroupPath))
```